[#100824938] Firewall core components for AWS

RichardKnop opened the pull request:
# What

Prior to any security testing and now that we understand the architecture better, we should configure firewalls to ensure that only whitelisted components/machines can talk to each other. All routes should be blacklisted by default and only opened if required.
# How this PR should be reviewed

This PR has been written to be reviewed in the following context:

I want to restrict access within the infrastructure and limit only ssh traffic by default

I want to allow the machines with in the private network to go out and get their updates

I want to allow the following access rules:
- Routers -> Docker Nodes on non privileged ports
- All internal machines to log their telegraf data to influxdb
- Gandalf -> Tsuru API on port 80
- Gandalf and Tsuru API -> Mongodb on port 27107
- API and Routers -> Redis on port 6379
- API, Docker nodes and Postgres -> Postgres on port 5432
- API and Gandalf -> Docker node on port 4243
- API -> Docker nodes to run healthchecks on non privileged ports
- Docker nodes -> Docker Registry on port 6000
- Docker nodes -> Elasticsearch on port 9200
# How to test this PR

To test this PR you can do the following:
- Create a new environment using this branch
- Do an ansible run and the run a smoketest of the environment
- Check that dashboard app is running
- Deploy a new app using any of the example applications available e.g. example-java-jetty and verify that it is running
- Use tsuru app-deploy to deploy an application
- Connect to the postgresql master and slave and verify that they are in sync with the same databases
- Connect to the grafana dashboard and verify that data is coming in

The examples above are just a sample of things you could try, please feel free to try more tests e.g. internal connectivity using nc or nmap.
# Who should review this PR

Anybody except Martin.

RichardKnop committed:
Removed default security group from Mongo/Redis instance

RichardKnop committed:
Specified from port directly (was zero in many places)

RichardKnop committed:
Fixed formatting. Removed ingres port 80 from tsuru api security group.

RichardKnop committed:
Fixed last two issues (api internal elb security group and duplicate rule)

RichardKnop committed:
Had to allow wider range of ports from router to docker node, otherwise was getting 502 from all apps

RichardKnop committed:
Allowed inbound traffic to InfluxDB on port 8086

RichardKnop commented:
Rebased against master.

RichardKnop committed:
Use default security group for inbound influxdb

All the nodes running telegraf write to influxdb db in port 8086

keymon commented on aws/coreos-docker-servers.tf:
```diff
@@ -30,3 +33,23 @@ resource "template_file" "etcd_discovery_url" {
   }
 }
+
+resource "aws_security_group" "docker_node" {
```
I wonder if it is worth it to separate the Docker Daemon traffic (port 4243) here. 

Only tsuru API nodes need to talk to the docker daemon, so we could block 4243 for everything but the docker api.

RichardKnop replied:
I'm not sure it can be separated. How would you do it? You mean restricting router to only 1024-4242 and 4244-65535?

keymon replied:
Yes, for instance. 

But I am not sure if it is worth it, because will prevent access from the router to the docker node. On one hand makes lots of sense, as the docker daemon is the most critical service there and if somebody manages to trick the router to hit the docker node we are gone. 

But on the other hand maybe we should secure the docker node in a different way, because a container could access the local docker daemon anyway. A more secure approach can be for instance use a different network interface for docker nodes, use ssh tunnels, or similar.

dcarley replied:
I think we should block it. Even though the implementation is kind of ugly and it doesn't cover container -> daemon.

keymon replied:
After discussing it in the team: We will block also the access to port 4243 for any host except the API nodes. 

Security groups won't prevent the app container accessing the local docker daemon, as the communication happens in the same host, but this will be tackled in a different story.

keymon commented:
I hit this bug (which I opened now) in this story: https://www.pivotaltracker.com/story/show/101590748 , but I don't think it is related to the PR.

RichardKnop committed:
Restrict 4243 access to only Tsuru API, no routers

keymon commented:
We still can connect from any host to the port 22 of any other host.

I think we shouldn't allow that, but instead we should only allow SSH traffic from the NAT server.

Example:
```
$ tsuru app-run -a postgresapi -- nc -v -w 1 <gandalfip> 22
Connection to 10.128.10.113 22 port [tcp/ssh] succeeded!
SSH-2.0-OpenSSH_6.6.1p1 Ubuntu-2ubuntu2
```

keymon commented on aws/security-groups.tf:
```diff
@@ -1,32 +1,17 @@
+    from_port       = 22
+    to_port         = 22
+    protocol        = "tcp"
+    self            = true
```
(as commented) I think we should allow only from NAT server for the internal hosts

RichardKnop replied:
Yes. I am getting some cycle errors though when I try to restrict this :(

keymon replied:
Maybe you should have 3 different rules: 
- For the nat server
- for gandalf  (must allow SSH for git)
- for the rest of servers

keymon commented on aws/gandalf.tf:
```diff
@@ -3,10 +3,55 @@ resource "aws_instance" "gandalf" {
 
+resource "aws_security_group" "gandalf" {
+  name = "${var.env}-tsuru-gandalf"
+  description = "Security group for Gandalf instance that allows SSH access from internet"
```
This description is outdated. The security group allows outbound traffic, SSH access from anyware, 8080 from api hosts, and 3232 (archive) from docker and api.

Alternatively we can split the group for each service: SSH, gandalf and archive-server.

mtekel commented:
I think all these findings should be reflected in GCE as well. Plus all the legacy rules/groups and descriptions cleaned up.

keymon commented on aws/api-servers.tf:
```diff
@@ -53,3 +61,41 @@ resource "aws_elb" "api-int" {
+    to_port         = 443
+    protocol        = "tcp"
+    security_groups = [
+      "${aws_security_group.web.id}"
```
The following issue does not come from the current PR but from before, but maybe we should takle it in the future.

Currently we have a [unique security group](https://github.com/alphagov/tsuru-terraform/pull/99/files#diff-c2eb5021b3f76e661aa1d2440598a977R27) for the API loadbalancers and the Routers ones. This means that the router ELB could reach the API if the node is added.

This is not a problem meanwhile we set the ELB backends in terraform, but we are considering autoregister the routers dynamically using something like [elb-presence](https://github.com/coreos/elb-presence). 

If we do so, we should split the `web` security group and create one for `api` and other for `router`.

keymon commented:
I added some more comments related to the commits themselves, but not about functionality. I am not sure if these comments should be a blocker to merge.

The PR keeps the functionality of the smoke test and the different platform features. I am currently double checking it blocks any other non authorised traffic.

RichardKnop commented:
@keymon Before the default security group allowed all inbound traffic which it doesn't anymore. So I have to add rule to NAT security group in order to allow outbound traffic from instances.

keymon committed:
Create dedicated security group for tsuru API

keymon committed:
Create dedicated security group for docker node

It will allow access to docker from  tsuru_api and router hosts.

keymon committed:
Dedicated security group for docker registry

Allow access to registry port 6000 from the  docker nodes.

keymon committed:
Move gandalf `sg` to gandalf terraform file

keymon committed:
Add dedicated security group for redis db

Allow traffic to redis port 6379 from tsuru api and routers.
Redis is use to define routes hipache, which are populated by tsuru api and read by the router.

keymon committed:
Create dedicated security group for mongodb

Allow access to mongodb to port 27017 for tsuru api and gandalf. Both services use mongodb as backend.

keymon committed:
Create dedicated security group for the routers

This will allow fine tune the security later on these hosts.

keymon committed:
Define dedicated security groups for API LBs

We only allow access to port 443 to api tsuru node from the API ELB instances.

keymon committed:
Add tags and env prefix to some security groups

In order to keep consistency in all the configuration, we add the environment prefixes 
and tags to all the defined security groups.

keymon committed:
Add dedicated security group for elastic search

keymon committed:
Move grafana security group to influx tf file

In order to keep all the related resources in the same files.

keymon committed:
Add dedicated security group for postgres

Allow access to postgres DB port 5432 only from docker nodes.

keymon committed:
Add dedicated security group for router LBs

A dedicated security group allows us restrict the access to the routers web ports to allow only the ELB ips.

keymon committed:
Move nat security group to nat tf file

In order to keep related objects in the same files.

keymon committed:
Move 'allow outbound' traffic from default to nat

Only nat server must allow all ingress traffic to route it out. Because 
that, we move the ingress accept all rule from `default` security group to `nat`
security group, so default only allows incoming SSH traffic.

keymon committed:
Format code in gandalf.tf for consistency

keymon committed:
Add default security group to db servers

This way we allow access to SSH.

keymon committed:
Allowed access to internal API ELB from Gandalf

Gandalf needs to access the tsuru api for authentication while pushing code.

keymon committed:
Allowed traffic to archive server from tsuru API

tsuru API needs to access the gandalf archive server in port 3232 while deployment.

keymon committed:
Allow API connect to any port in docker nodes

Tsuru API needs to check the health of the units during deployment or scaling, and connects directly.

keymon committed:
Allow traffic to archive server from docker nodes

While deployment, the docker nodes need to download the application binaries from the gandalf archive server in port 3232.

keymon committed:
Use web security group for api&router ELBs

We use a common security group for all the ELBs listening in ports 80&443, as the rules are the same.
Access from the ELBs to the backend is restricted by the ELB backend definition.

keymon commented:
We split some of the commits for more detailed documentation. We will create a different PR for secure the SSH in all hosts, and only allow access from NAT server.

keymon commented:
Visualisation of this setup:

![viz](https://cloud.githubusercontent.com/assets/280032/9380473/52a93ee2-4728-11e5-884f-1361a9ebe8c3.png)
